EventListener/ReportSubscriber.php

```diff
@@ -0,0 +1,268 @@
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+class ReportSubscriber implements EventSubscriberInterface
+{
+    public const CONTEXT_COMPANY_SEGMENTS         = 'company_segments';
+    public const COMPANY_TABLE                    = 'companies';
+    public const COMPANIES_PREFIX                 = 'comp';
+    public const COMPANY_SEGMENTS_XREF_PREFIX     = 'csx';
+    public const COMPANY_SEGMENTS_XREF_TABLE      = 'company_segment_xref';
+
+    public function __construct(
+        private CompanyReportData $companyReportData,
+        private CompanySegmentRepository $companySegmentRepository,
+        private Connection $db,
+    ) {
+    }
+
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            ReportEvents::REPORT_ON_BUILD    => ['onReportBuilder', 0],
+            ReportEvents::REPORT_ON_GENERATE => ['onReportGenerate', 0],
+        ];
+    }
+
+    public function onReportBuilder(ReportBuilderEvent $event): void
+    {
+        if (!$event->checkContext([self::CONTEXT_COMPANY_SEGMENTS])) {
+            return;
+        }
+
+        $columns = $this->companyReportData->getCompanyData();
+        unset($columns['companies_lead.is_primary'], $columns['companies_lead.date_added']);
+
+        $segmentList = $this->getFilterSegments();
+
+        $segmentFilter = [self::COMPANY_SEGMENTS_XREF_PREFIX.'.segment_id' => [
+            'alias'     => 'companysegments',
+            'label'     => 'mautic.company_segments.report.company_segments',
+            'type'      => 'select',
+            'list'      => $segmentList,
+            'operators' => [
+                'in'       => 'mautic.core.operator.in',
+                'notIn'    => 'mautic.core.operator.notin',
+                'empty'    => 'mautic.core.operator.isempty',
+                'notEmpty' => 'mautic.core.operator.isnotempty',
+            ],
+        ],
+        ];
+        $filters = array_merge($columns, $segmentFilter);
+
+        $event->addTable(
+            self::CONTEXT_COMPANY_SEGMENTS,
+            [
+                'display_name' => 'mautic.company_segments.report.company_segments',
+                'columns'      => $columns,
+                'filters'      => $filters,
+            ],
+            'companies'
+        );
+
+        $event->addGraph(self::CONTEXT_COMPANY_SEGMENTS, 'line', 'mautic.lead.graph.line.companies');
+        $event->addGraph(self::CONTEXT_COMPANY_SEGMENTS, 'pie', 'mautic.lead.graph.pie.companies.industry');
+        $event->addGraph(self::CONTEXT_COMPANY_SEGMENTS, 'pie', 'mautic.lead.table.pie.company.country');
+        $event->addGraph(self::CONTEXT_COMPANY_SEGMENTS, 'table', 'mautic.lead.company.table.top.cities');
+    }
+
+    /**
+     * @return array<string, string>
+     */
+    public function getFilterSegments(): array
+    {
+        $segments    = $this->companySegmentRepository->getSegmentObjectsViaListOfIDs([]);
```

Comment:
Please return what you need from the repository directly. No need to fetch the whole object to have id=>name.